### PR TITLE
Schema source refactoring

### PR DESCRIPTION
**Pull request by pbetkier**
Refactoring which introduces a level of indirection when getting message schema for topic. Allows introducing a different schema source (e.g. Schema Repo service) in the future.

### CODE REVIEW

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/converter/AvroToJsonConverter.java [hunk 1]
```diff
@@ -14,32 +14,22 @@
...
-    private final Schema schema;
-
-    public AvroToJsonConverter(Schema schema) {
-        this.schema = schema;
-        this.reader = new GenericDatumReader<>(schema);
-        this.writer = new GenericDatumWriter<>(schema);
-    }
-
-    public byte [] convert(byte [] data) {
+    public byte[] convert(byte[] data, Schema schema) {
```
Comment (pawelszymczyk) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/converter/AvroToJsonConverter.java]: this method could be static now

File: hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/AvroSchemaRepositoryFactory.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    private final SchemaSourceProvider schemaSourceProvider;
+
+    @Inject
+    public AvroSchemaRepositoryFactory(SchemaSourceProvider schemaSourceProvider) {
+        this.schemaSourceProvider = schemaSourceProvider;
+    }
+
+    @Override
+    public SchemaRepository<Schema> provide() {
+        return new SchemaRepository<>(schemaSourceProvider, Executors.newFixedThreadPool(2), source -> new Schema.Parser().parse(source.value()));
```
Comment (pawelszymczyk) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/AvroSchemaRepositoryFactory.java]: thread pool size should be configurable

File: hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/CouldNotCompileSchemaException.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package pl.allegro.tech.hermes.domain.topic.schema;
+
+public class CouldNotCompileSchemaException extends RuntimeException {
```
Comment (pawelszymczyk) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/CouldNotCompileSchemaException.java]: I'm suggest to follow our convention and extends HermesException

File: hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/JsonSchemaRepositoryFactory.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Inject
+    public JsonSchemaRepositoryFactory(ObjectMapper objectMapper, SchemaSourceProvider schemaSourceProvider) {
+        this.objectMapper = objectMapper;
+        this.schemaSourceProvider = schemaSourceProvider;
+    }
+
+    @Override
+    public SchemaRepository<JsonSchema> provide() {
+        jsonSchemaFactory = JsonSchemaFactory.byDefault();
```
Comment (pawelszymczyk) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/JsonSchemaRepositoryFactory.java]: maybe create JsonSchemaFactory in constructor and set variable as final?

File: hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/JsonSchemaRepositoryFactory.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        return new SchemaRepository<>(schemaSourceProvider, Executors.newFixedThreadPool(2),
+                source -> {
+                    try {
+                        return jsonSchemaFactory.getJsonSchema(objectMapper.readTree(source.value()));
+                    } catch (IOException | ProcessingException e) {
+                        throw new RuntimeException(e);
```
Comment (pawelszymczyk) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/JsonSchemaRepositoryFactory.java]: probably it should be transformed to our exception (extending HermesException)

File: hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/SchemaRepository.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    SchemaRepository(SchemaSourceProvider schemaSourceProvider, ExecutorService reloadSchemaSourceExecutor, Ticker ticker, SchemaCompiler<T> schemaCompiler) {
+        this.schemaCompiler = schemaCompiler;
+        this.schemaCache = CacheBuilder
+                .newBuilder()
+                .ticker(ticker)
+                .refreshAfterWrite(10, TimeUnit.MINUTES)
+                .expireAfterWrite(24, TimeUnit.HOURS)
+                .build(new CacheLoader<Topic, SchemaWithSource>() {
```
Comment (pawelszymczyk) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/SchemaRepository.java]: i think it should be extracted to another place eg private factory method or external class

File: hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/SchemaRepository.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+                                logger.warn("Could not compile schema for topic {}", topic.getQualifiedName(), e);
+                                throw e;
+                            }
+                        });
+                        reloadSchemaSourceExecutor.execute(task);
+                        return task;
+                    }
+
+                    private SchemaSource findSchemaSource(Topic topic) {
+                        return schemaSourceProvider.get(topic).orElseThrow(() -> new SchemaSourceNotFoundException("No schema source for topic " + topic.getQualifiedName()));
```
Comment (pawelszymczyk) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/SchemaRepository.java]: you could extract "No schema source for topic " to SchemaSourceNotFoundException internals and pass only qualifiedTopicName

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/api/SchemaEndpoint.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        Topic topic = topicService.getTopicDetails(fromQualifiedName(qualifiedTopicName));
+        Optional<SchemaSource> schemaSource = schemaSourceRepository.get(topic);
+        return schemaSource.map(SchemaSource::value).map(v -> Response.ok(v).build()).orElse(Response.noContent().build());
+    }
+
+    @POST
+    @Consumes(APPLICATION_JSON)
+    @RolesAllowed({Roles.GROUP_OWNER, Roles.ADMIN})
+    @ApiOperation(value = "Save schema", httpMethod = HttpMethod.POST)
+    public Response save(@PathParam("topicName") String qualifiedTopicName, String schema) {
```
Comment (pawelszymczyk) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/api/SchemaEndpoint.java]: what about update schema?

**Comment by pawelszymczyk** [comment]
+:100:

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/api/SchemaEndpoint.java (hunk 8 above)
Comment (dankraw) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/api/SchemaEndpoint.java]: as already discussed we're gonna keep it as is 
:dancers:

File: hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/SchemaRepository.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    public SchemaRepository(SchemaSourceProvider schemaRepository, ExecutorService reloadSchemaSourceExecutor, SchemaCompiler<T> schemaCompiler) {
+        this(schemaRepository, reloadSchemaSourceExecutor, Ticker.systemTicker(), schemaCompiler);
+    }
+
+    SchemaRepository(SchemaSourceProvider schemaSourceProvider, ExecutorService reloadSchemaSourceExecutor, Ticker ticker, SchemaCompiler<T> schemaCompiler) {
+        this.schemaCompiler = schemaCompiler;
+        this.schemaCache = CacheBuilder
+                .newBuilder()
+                .ticker(ticker)
+                .refreshAfterWrite(10, TimeUnit.MINUTES)
```
Comment (pjagielski) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/SchemaRepository.java]: maybe this could be configurable?

**Comment by pjagielski** [comment]
:+1:

**Pull request merged**